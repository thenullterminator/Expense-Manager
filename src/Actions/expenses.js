import uuid from 'uuid';

// Action Generators for expense reducers.
const addExpense=({description='',note='',amount=0,createdAt=0}={})=>{

      return {
                   type:'ADD_EXPENSE',
                   expense: {
                        id:uuid(),
                        description,
                        note,
                        amount,
                        createdAt
             }
       };
 };
 
 const removeExpense=({id}={})=>{
 
       return{
                   type:'REMOVE_EXPENSE',
                   id
        };
 };
 
 const editExpense=(id,updates)=>{
       return{
             type:'EDIT_EXPENSE',
             id,
             updates
       };
 
 };

export {addExpense,removeExpense,editExpense};