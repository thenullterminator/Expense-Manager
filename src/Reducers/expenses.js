// Expense Reducer
const expenseDefaultState=[];
const expenseReducer=(state=expenseDefaultState,action)=>{

      switch(action.type)
      {
            case 'ADD_EXPENSE':
                  return [...state,action.expense];
            case 'REMOVE_EXPENSE':
                  return state.filter(({id})=>id!==action.id);
            case 'EDIT_EXPENSE' :
                  return state.map((expense)=>{
                        if(expense.id===action.id)
                        {
                              return{
                                    ...expense,
                                    ...action.updates
                              };
                        }
                        else
                        {
                              return expense;
                        }
                  });
            default:
                  return state;
      }

};

export default expenseReducer;