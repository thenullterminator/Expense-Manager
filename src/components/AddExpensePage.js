import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import {addExpense} from '../Actions/expenses';


 export class AddExpensePage extends React.Component{

      onSubmit=(expense)=>{
            this.props.onSubmit(expense);
            this.props.history.push('/');//Redirecting to home page.
      };

      render(){
            return (

            <div>
                  <h2>Add Expense</h2>
                  <ExpenseForm onSubmit={this.onSubmit}/>
            </div>

            );
      }
}


const mapDispatchToProps=(dispatch)=>{
      return {
            onSubmit:(expense)=>dispatch(addExpense(expense))
      };
};

export default connect(undefined,mapDispatchToProps)(AddExpensePage);
