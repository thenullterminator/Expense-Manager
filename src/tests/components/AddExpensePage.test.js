import React from 'react';
import {shallow} from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage.js';
import expenses from '../fixtures/expenses';

let onSubmit,wrapper,history;

beforeEach(()=>{
      onSubmit=jest.fn();
      history={push:jest.fn()};
      wrapper=shallow(<AddExpensePage onSubmit={onSubmit} history={history}/>);
});

test('should render Add expense page correctly',()=>{
      expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit correctly',()=>{

      wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
      expect(history.push).toHaveBeenCalledWith('/');
      expect(onSubmit).toHaveBeenCalledWith(expenses[1]);
});