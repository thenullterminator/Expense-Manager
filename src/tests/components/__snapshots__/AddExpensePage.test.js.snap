// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Add expense page correctly 1`] = `
<div>
  <h2>
    Add Expense
  </h2>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
